File: sections/35_math_mixed_contest/problems/k_weighted_coupon_collection/solution.py
import sys


MOD = 1_000_000_007


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    n = data[0]
    weight = data[1:]
    total = sum(weight)

    subset_sum = [0] * (1 << n)
    answer = 0
    for mask in range(1, 1 << n):
        least_bit = mask & -mask
        bit = least_bit.bit_length() - 1
        subset_sum[mask] = subset_sum[mask ^ least_bit] + weight[bit]
        term = total * pow(subset_sum[mask], MOD - 2, MOD) % MOD
        answer += term if mask.bit_count() & 1 else -term

    print(answer % MOD)


if __name__ == "__main__":
    main()

File: sections/35_math_mixed_contest/problems/k_weighted_coupon_collection/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using int64 = long long;
const int64 MOD = 1'000'000'007;

int64 power(int64 base, int64 exponent) {
    int64 result = 1;
    while (exponent) {
        if (exponent & 1)
            result = result * base % MOD;
        base = base * base % MOD;
        exponent >>= 1;
    }
    return result;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<int64> weight(n);
    int64 total = 0;
    for (int64& value : weight) {
        cin >> value;
        total += value;
    }

    int subset_count = 1 << n;
    vector<int64> subset_sum(subset_count);
    int64 answer = 0;
    for (int mask = 1; mask < subset_count; ++mask) {
        int bit = __builtin_ctz(mask);
        subset_sum[mask] = subset_sum[mask ^ (1 << bit)] + weight[bit];
        int64 term = total * power(subset_sum[mask], MOD - 2) % MOD;
        if (__builtin_popcount(static_cast<unsigned>(mask)) & 1)
            answer = (answer + term) % MOD;
        else
            answer = (answer - term + MOD) % MOD;
    }

    cout << answer << '\n';
}
